File: pyinterpreter/interpreter/interpreter.py
from .ast import Number, BinOp, UnOp
from interpreter.parser import Parser


class Interpreter:
    def __init__(self):
        self.parser = Parser()

    def visit(self, node):
        if isinstance(node, Number):
            return self.visit_number(node)
        elif isinstance(node, BinOp):
            return self.visit_binop(node)
        elif isinstance(node, UnOp):
            return self.visit_inop(node)

    def visit_number(self, node):
        return float(node.token.value)

    def visit_binop(self, node):
        match node.op.value:
            case "+":
                return self.visit(node.left) + self.visit(node.right)
            case "-":
                return self.visit(node.left) - self.visit(node.right)
            case "*":
                return self.visit(node.left) * self.visit(node.right)
            case "/":
                return self.visit(node.left) / self.visit(node.right)

    def visit_inop(self, node):
        match node.op.value:
            case "+":
                return +self.visit(node.value)
            case "-":
                return -self.visit(node.value)
            case _:
                raise ValueError("Invalid operator")

    def eval(self, code):
        tree = self.parser.parse(code)
        return self.visit(tree)

File: pascalInterpreter/src/main/kotlin/interpreter/Interpreter.kt
package interpreter

import Exceptions.SyntaxException

class Interpreter {
    private val parser = Parser()
    val variables = HashMap<String, Float>()

    fun visit(node: Node): Float {
        when (node) {
            is PascalNumber -> return visitNumber(node)
            is BinOp -> return visitBinOp(node)
            is Variable -> visitVariable(node)
            is Semi -> visitSemi(node)
            is Empty -> {}
            is Identifier -> return visitIdentifier(node)
            else -> return visitUnOp(node as UnOp)
        }
        return 0.0f
    }

    fun visitIdentifier(node: Identifier): Float {
        if (node.token.value in variables.keys) {
            return variables[node.token.value]!!
        }
        throw SyntaxException("Unknown variable: ${node.token.value}")
    }

    fun visitSemi(node: Semi) {
        visit(node.left)
        visit(node.right)
    }

    fun visitVariable(node: Variable) {
        if (node.token.value !in variables.keys) {
            variables[node.token.value] = 0.0f
        }
        variables[node.token.value] = visit(node.value)
    }

    fun visitNumber(number: PascalNumber): Float = number.token.value.toFloat()

    fun visitBinOp(node: BinOp): Float {
        return if (node.op.value == "+") {
            visit(node.left) + visit(node.right)
        } else if (node.op.value == "-") {
            visit(node.left) - visit(node.right)
        } else if (node.op.value == "*") {
            visit(node.left) * visit(node.right)
        } else {
            visit(node.left) / visit(node.right)
        }

//        when (node.op.value) {
//            "+" -> return visit(node.left) + visit(node.right)
//            "-" -> return visit(node.left) - visit(node.right)
//            "*" -> return visit(node.left) * visit(node.right)
//            else -> return visit(node.left) / visit(node.right)
//        }
    }

    fun visitUnOp(node: UnOp): Float {
        return when (node.op.value) {
            "+" -> +visit(node.value)
            "-" -> -visit(node.value)
            else -> throw SyntaxException("Invalid operator")
        }
    }

    fun eval(code: String): Float {
        val tree = parser.parse(code)
        return visit(tree)
    }
}